### DISCUSSION TITLE
Resource limitations: Enforcing pagination limits at router level

### DISCUSSION THREAD

**Discussion by andrew-kolesnikov** [apollographql/router#1246] [Q&A] [answered]
What is a good way to enforce pagination limits at the router level, without delegating to subgraphs?

Here's an example:
`query Comments($cursor: String) { Comments(first: 99999999999999999, after: $cursor) { ...`

In other words, I am looking to implement a basic version of https://docs.github.com/en/graphql/overview/resource-limitations at the router level

**Comment by garypen** [comment, marked as answer]
I think the best way to address this would be to [write a plugin](https://www.apollographql.com/docs/router/customizations/overview/). You've got two options: depending on how sophisticated you want your processing to be. If it's very simple, then you might be able to achieve your goal with a rhai plugin, which facilitates basic text manipulation, math, etc.. However, I think you are looking for something more sophisticated, since the link you provide seems to require knowledge of GraphQL Query structure. This would probably require you to write a rust plugin.

You may be able to get the information you need to make decisions from the `QueryPlan` that we provide in `query_planning_service`. If not, you will need to parse the request (perhaps using [apollo-parser](https://crates.io/crates/apollo-parser)) and then process the resulting SyntaxTree to extract the required information and implement your controls. You would access the request using `map_request`.

I don't think this is a trivial piece of work.
